W-11984725-logger-revamp-duke (#2402)
mulesoft/docs-mule-runtime-3: modules/ROOT/pages/logger-component-reference.adoc

					<!-- Transform Message -->
					<ee:transform doc:name="Transform Message" >
						<ee:message>
							<ee:set-payload><![CDATA[%dw 2.0
output application/json
---
(payload + 100)]]></ee:set-payload>
						</ee:message>
					</ee:transform>
					<!-- Third Logger -->
					<logger level="INFO" doc:name="Logger 3"
					        message="#[payload]"
					        category="AGGREGATOR: THIRD LOGGER" />
				</batch:aggregator>
			</batch:step>
		</batch:process-records>
	</batch:job>
</flow>
----

* Logger 1, the first Logger component within the Aggregator, receives the array as a non-repeatable stream (`stream`), so it prints `[1, 2]` but passes an empty array `[]` to the next component in the Aggregator.
* The first Transform Message component receives the empty array from Logger 1 and sends it to Logger 2. So Logger 2 prints an _empty_ array (`[ ]`).
* The second Transform Message component appends `100` to the empty array, so Logger 3 prints the new payload, `[100]`.

The Loggers print the following (edited for readability):

[source,logs]
----
INFO  ...AGGREGATOR: FIRST LOGGER: [1, 2]
INFO  ...AGGREGATOR: SECOND LOGGER: [
]
INFO  ... AGGREGATOR: THIRD LOGGER: [
  100
]
----

> Assume a scenario in which the first Transform Message component is located before the Logger 1. In this case, the Logger prints the following (edited):

Author: ```suggestion
Unlike the Logger, a Transform Message, converts its input into a non-streamed form. So a Logger located downstream of such a component does not receive a streamed payload.

Assume a scenario in which the first Transform Message component from the <<example, preceding example>> is located _before_ the Logger 1. In this case, the Logger prints the following (edited):
```

MyCustomCategory: My Message
----

The following example places a DataWeave selector expression within a `message` string:

[source,xml]
----
<logger category="monitoring" message="Message #[payload.id] processed successfully" level="INFO"/>
----

Assuming a payload such as `{"id" : "100", "some-key" : "some value" }`, the Logger prints the message `Message 100 processed successfully`, for example:

[source,xml]
----
INFO  2022-11-10 21:11:07,791 [[MuleRuntime].uber.09:
[w-logger].logger-example-Flow.CPU_LITE @600dcddb]
[processor: logger-example-Flow/processors/1;
event: 3fd13710-617f-11ed-900f-147ddaaf4f97]
org.mule.runtime.core.internal.processor.LoggerMessageProcessor:
Message 100 processed successfully
----

This Logger records the processing time of a flow, using a single expression and DataWeave concatenation:

//TODO: Need to show how to get processing time or use a better example that doesn't throw an error.
[source,xml]
----
<logger category="performance" message="#['Message ' ++ payload.id ++ ' took ' ++ vars.processingTime ++ ' milliseconds to process']" level="INFO"/>
----

[[non_repeatable_streams]]
== Non-Repeatable Streams

> In Mule, a non-repeatable stream can be read only once. The Logger consumes a payload that is streamed in a non-repeatable way and does not pass it to the next component in the flow.

Author: ```suggestion
In Mule, a non-repeatable stream can be read only once. The Logger consumes a payload that is streamed in a non-repeatable way and does not pass it to the next component in the flow. In the next component, the `payload` selector returns a `null` or empty value. 
```

> This scenario can occur if a Mule components, such as as a xref:file-connector::file-documentation.adoc#write[File Write operation] or the xref:batch-reference.adoc#ref_batch_aggregator_component[Batch Aggregator] component, is configured with a _non-repeatable_ streaming strategy (see xref:tuning-repeatable-nonrepeatable.adoc[]).

Author: ```suggestion
This scenario can occur if a Mule component, such as as a xref:file-connector::file-documentation.adoc#write[File Write operation] or the xref:batch-reference.adoc#ref_batch_aggregator_component[Batch Aggregator] component, is configured with a _non-repeatable_ streaming strategy (see xref:tuning-repeatable-nonrepeatable.adoc[]).
```

> A component, such as Transform Message, converts such a payload into a non-streamed form. So a Logger located downstream of such a component no longer receives streamed payload.

Author: ```suggestion
Unlike the Logger, many Mule components, such as Transform Message, convert such a payload into a non-streamed form. So a Logger located downstream of such a component no longer receives streamed payload.
```

> Unlike the Logger, many Mule components, such as Transform Message, convert such a payload into a non-streamed form. So a Logger located downstream of such a component no longer receives streamed payload.

Author: ```suggestion
```

MyCustomCategory: My Message
----

The following example places a DataWeave selector expression within a `message` string:

[source,xml]
----
<logger category="monitoring" message="Message #[payload.id] processed successfully" level="INFO"/>
----

Assuming a payload such as `{"id" : "100", "some-key" : "some value" }`, the Logger prints the message `Message 100 processed successfully`, for example:

[source,xml]
----
INFO  2022-11-10 21:11:07,791 [[MuleRuntime].uber.09:
[w-logger].logger-example-Flow.CPU_LITE @600dcddb]
[processor: logger-example-Flow/processors/1;
event: 3fd13710-617f-11ed-900f-147ddaaf4f97]
org.mule.runtime.core.internal.processor.LoggerMessageProcessor:
Message 100 processed successfully
----

This Logger records the processing time of a flow, using a single expression and DataWeave concatenation:

//TODO: Need to show how to get processing time or use a better example that doesn't throw an error.
[source,xml]
----
<logger category="performance" message="#['Message ' ++ payload.id ++ ' took ' ++ vars.processingTime ++ ' milliseconds to process']" level="INFO"/>
----

[[non_repeatable_streams]]
== Non-Repeatable Streams

> In Mule, a non-repeatable stream can be read only once. Unlike most other Mule components, the Logger does not convert such a stream to a non-streamed or repeatable form.

Author: ```suggestion
In Mule, a non-repeatable stream can be read only once. The Logger consumes a payload that is streamed in a non-repeatable way and does not pass it to the next component in the flow.
```

> If a Logger receives a Mule event with a payload that is configured as a non-repeatable stream, the Logger consumes the payload, so the payload received by the next component in the flow returns `null`.

Author: ```suggestion
```

This scenario can occur if a Mule component, such as as a xref:file-connector::file-documentation.adoc#write[File Write operation] or the xref:batch-reference.adoc#ref_batch_aggregator_component[Batch Aggregator] component, is configured with a _non-repeatable_ streaming strategy (see xref:tuning-repeatable-nonrepeatable.adoc[]).

> To illustrate this scenario, assume a Batch Aggregator that contains two Logger components is set to stream (`stream`) it input, the array `[1,2]`:

Author: ```suggestion
[[example]]
To illustrate this scenario, assume a Batch Aggregator that contains two Logger components is set to stream (`stream`) it input, the array `[1,2]`:
```

endif::[]
:keywords: mule, esb, studio, logger, logs, log, notifications, errors, debug

> Logs help you monitor and debug your Mule applications by printing important information such as error messages, status notifications, payload data, and so on. The Logger component (`<logger/>`) provides a way to add custom log messages to the logs. You can configure any number of Logger components in a flow to print a string, the output of a DataWeave expression, or any combination of strings and expressions.

Author: ```suggestion
Logs help you monitor and debug your Mule applications by printing important information such as error messages, status notifications, payload data, and so on. The Logger component (`<logger/>`) provides a way to add custom log messages to the logs. You can configure any number of Logger components in a flow with a string, a DataWeave expression, or any combination of strings and expressions.
```

  <logger message="#[payload]" level="INFO"/>
</flow>
----

In the Studio console, the logged message, `Hello MuleSoft!` looks like this:

[source,logs]
----
INFO  2022-11-10 19:39:18,222 [[MuleRuntime].uber.01: [w-logger].logger-example-Flow.CPU_LITE @6f862394]
[processor: logger-example-Flow/processors/1; event: 6bd2b350-6172-11ed-9f87-147ddaaf4f97]
org.mule.runtime.core.internal.processor.LoggerMessageProcessor: Hello MuleSoft!
----

* The configurable log level is `INFO`. See <<reference>>.
* A date-time stamp `2022-11-10 19:39:18,222` marks an execution of a Logger.
* CPU_LITE is the See xref:execution-engine.adoc#processing_types[processing type].
* `event: 6bd2b350-6172-11ed-9f87-147ddaaf4f97` identifies the Mule event to which the logged message applies.
* The message `LoggerMessageProcessor` processor prints the message `Hello MuleSoft!`.

[[reference]]
== Reference

The Logger provides configurable properties for the name of the logger, the message to log, the log level, and a string you can use to name or categorize a type of log message.

image::logger.png[]

[%header%autowidth.spread,cols="1a,1a,4a"]
|===
| Field | Attribute | Description

> | *Message* | `message` | String or DataWeave expression that specifies the Mule log message. By default, messages log to the application's log file. Example:

Author: ```suggestion
| *Message* | `message` | String or DataWeave expression that specifies the Mule log message. By default, messages log to the application's log file and in the console where the application runs, such as the Anypoint Studio console or the OS console for on-premises deployments. Example:
```

MyCustomCategory: My Message
----

The following example places a DataWeave selector expression within a `message` string:

[source,xml]
----
<logger category="monitoring" message="Message #[payload.id] processed successfully" level="INFO"/>
----

Assuming a payload such as `{"id" : "100", "some-key" : "some value" }`, the Logger prints the message `Message 100 processed successfully`, for example:

[source,xml]
----
INFO  2022-11-10 21:11:07,791 [[MuleRuntime].uber.09:
[w-logger].logger-example-Flow.CPU_LITE @600dcddb]
[processor: logger-example-Flow/processors/1;
event: 3fd13710-617f-11ed-900f-147ddaaf4f97]
org.mule.runtime.core.internal.processor.LoggerMessageProcessor:
Message 100 processed successfully
----

This Logger records the processing time of a flow, using a single expression and DataWeave concatenation:

//TODO: Need to show how to get processing time or use a better example that doesn't throw an error.
[source,xml]
----
<logger category="performance" message="#['Message ' ++ payload.id ++ ' took ' ++ vars.processingTime ++ ' milliseconds to process']" level="INFO"/>
----

[[non_repeatable_streams]]
== Non-Repeatable Streams

> In Mule, a non-repeatable stream can be read only once. The Logger consumes a payload that is streamed in a non-repeatable way and does not pass it to the next component in the flow. In the next component, the `payload` selector returns a `null` or empty value. 

Author: When Logger reads a value that's a non-repeatable stream, that value is consumed. 

Logger itself doesn't return anything (void). 

Transform Message is mutating the payload. Reads and applies a transformation.

Author: Output of Logger is void.

This scenario can occur if a Mule component, such as as a xref:file-connector::file-documentation.adoc#write[File Write operation] or the xref:batch-reference.adoc#ref_batch_aggregator_component[Batch Aggregator] component, is configured with a _non-repeatable_ streaming strategy (see xref:tuning-repeatable-nonrepeatable.adoc[]).

> [[example]]

Author: ```suggestion
[[example_non_repeatable_stream]]
```

						</ee:message>
					</ee:transform>
					<!-- Third Logger -->
					<logger level="INFO" doc:name="Logger 3"
					        message="#[payload]"
					        category="AGGREGATOR: THIRD LOGGER" />
				</batch:aggregator>
			</batch:step>
		</batch:process-records>
	</batch:job>
</flow>
----

* Logger 1, the first Logger component within the Aggregator, receives the array as a non-repeatable stream (`stream`), so it prints `[1, 2]` but passes an empty array `[]` to the next component in the Aggregator.
* The first Transform Message component receives the empty array from Logger 1 and sends it to Logger 2. So Logger 2 prints an _empty_ array (`[ ]`).
* The second Transform Message component appends `100` to the empty array, so Logger 3 prints the new payload, `[100]`.

The Loggers print the following (edited for readability):

[source,logs]
----
INFO  ...AGGREGATOR: FIRST LOGGER: [1, 2]
INFO  ...AGGREGATOR: SECOND LOGGER: [
]
INFO  ... AGGREGATOR: THIRD LOGGER: [
  100
]
----

Unlike the Logger, a Transform Message, converts its input into a non-streamed form. So a Logger located downstream of such a component does not receive a streamed payload.

> Assume a scenario in which the first Transform Message component from the <<example, preceding example>> is located _before_ the Logger 1. In this case, the Logger prints the following (edited):

Author: ```suggestion
Assume a scenario in which the first Transform Message component from the << example_non_repeatable_stream, preceding example>> is located _before_ the Logger 1. In this case, the Logger prints the following (edited):
```

This Core component helps you monitor and debug your Mule application by logging important information such as error messages, status notifications, payloads, and so on. You can add a Logger anywhere in a flow, and you can configure it to log a string that you specify, the output of a DataWeave expression you write, or any combination of strings and expressions.

[WARNING]
--

> Keep in mind that the Logger is one of the only components that supports mixing Strings and expressions within a value. DataWeave String interpolation or concatenation within a single expression should be used elsewhere.

Author: moved to reference

--
//TODO: IS THIS TRUE BEYOND ON-PREM INSTALLATIONS? 
The configured messages are logged in a Mule application's log file, which is located in `MULE_HOME/logs/<app-name>.log` if no custom log file path is specified in the `log4j2.xml` file. For details, see xref:logging-and-debugging.adoc[].

The configured messages are logged to the app's log file, which is located in `MULE_HOME/logs/<app-name>.log` if no custom log file path is specified in the `log4j2.xml` file. +
In Studio, the logs show in the Console.
In Anypoint Studio, the logs print to the xref:studio::index.adoc#console[console].

> The following example displays the message in Set Payload in a browser and also logs the message.

Author: removed bc "... in a browser": it's in the browser bc of the request to the listener, not directly relevant to logger

> The following example logs the message.

Author: ```suggestion
The following example logs the message `Hello MuleSoft!`.
```
